exist-core/src/main/java/org/exist/xquery/value/AtomicValueComparator.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (C) 2014, Evolved Binary Ltd
+ *
+ * This file was originally ported from FusionDB to eXist-db by
+ * Evolved Binary, for the benefit of the eXist-db Open Source community.
+ * Only the ported code as it appears in this file, at the time that
+ * it was contributed to eXist-db, was re-licensed under The GNU
+ * Lesser General Public License v2.1 only for use in eXist-db.
+ *
+ * This license grant applies only to a snapshot of the code as it
+ * appeared when ported, it does not offer or infer any rights to either
+ * updates of this source code or access to the original source code.
+ *
+ * The GNU Lesser General Public License v2.1 only license follows.
+ *
+ * ---------------------------------------------------------------------
+ *
+ * Copyright (C) 2014, Evolved Binary Ltd
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; version 2.1.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package org.exist.xquery.value;
+
+import com.ibm.icu.text.Collator;
+import net.jcip.annotations.ThreadSafe;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.exist.http.RESTServer;
+import org.exist.xquery.XPathException;
+
+import javax.annotation.Nullable;
+import java.util.Comparator;
+
+/**
+ * @author <a href="mailto:adam@evolvedbinary.com">Adam Retter</a>
+ */
+@ThreadSafe
+public class AtomicValueComparator implements Comparator<AtomicValue> {
+
+    protected static final Logger LOG = LogManager.getLogger(RESTServer.class);
+
+    private final @Nullable Collator collator;
+
+    public AtomicValueComparator() {
+        this(null);
+    }
+
+    public AtomicValueComparator(@Nullable final Collator collator) {
+        this.collator = collator;
+    }
+
+    @Override
+    public int compare(final AtomicValue o1, final AtomicValue o2) {
+        if (o1 == null && o2 == null) {
+            return 0;
+        } else if (o1 == null) {
+            return -1;
+        } else if (o2 == null) {
+            return 1;
+        }
+
+        try {
+            return o1.compareTo(collator, o2);
+        } catch (final XPathException e) {
+            LOG.error(e.getMessage(), e);
+            throw new ClassCastException(e.getMessage());
```

Comment:
??

Reply:
@dizzzz What other option do we have? The API says we can throw either:

```
     * @throws NullPointerException if an argument is null and this
     *         comparator does not permit null arguments
     * @throws ClassCastException if the arguments' types prevent them from
     *         being compared by this comparator.
```

So `ClassCastException` seemed like the *most* correct option to me.